### PR TITLE
Users can now upload an image to their comments

### PR DESCRIPTION
**Pull request by anamariamilcu**
Created another Servlet that does the connection between Blobstore and the page itself, providing a link for uploading an image. This URL is then stored in the DataStore in the Comment type entities, if the comment has an image. The link is then used to display the image in the comment section.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -75,11 +87,14 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
     Integer commentsNumber = defaultCommentsNumber;
     Integer commentsCount = 0;
-    try {
-      commentsNumber = Integer.parseInt(request.getParameter("comments_number"));
-    } catch (NumberFormatException e) {
-      // If the data provided by the user is invalid send a 400 Bad Request.
-      response.sendError(HttpServletResponse.SC_BAD_REQUEST, e.getMessage());
+    /* Avoid null pointer exception. */
```
Comment (alexcepoi) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:91]: All of this seems unrelated again :)

I guess what's happening here is as you're submitting more code you're realizing you're breaking the app. So you need to fix it, before you get to implement a new feature. In the next week, you're going to learn about test-driven development, which hopefully will ensure that your code will continue to work even as you're adding new features.

Tests can still fail you, so when you realize the app is broken, it's still a very good idea to submit a small PR to fix it. You will be surprised how often a 'fix' sometimes breaks stuff even further and needs to be `unfixed`, so generally putting fixes and new features in the same PR is a recipe for disaster (the new feature is not ready yet, but the fix has to go in ASAP).

A good mindset to get into is to always check your current diff (e.g. to diff against master `git diff master`). Whenever you think you need to fix something unrelated, you switch to master, make a new branch, create a new PR.

You can then rebase your feature branch on top of the fix branch (`git rebase --onto fix_branch feature_branch`).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -133,7 +149,53 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
```
Comment (alexcepoi) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:174]: Are you sure the user submitted at least one? Sure the form has a single file input, but what if someone tampered with that.

Remember to never trust the client.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -133,7 +149,53 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // Check the validity of the file here, to make sure it's an image file.
+    if (blobInfo.getContentType().contains("image") == false) {
```
Comment (alexcepoi) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Or even better .startsWith("image/")

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -133,7 +149,53 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Only images allowed!");
+      blobstoreService.delete(blobKey);
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // To support running in Google Cloud Shell with AppEngine's devserver, we must use the relative
+    // path to the image, rather than the path returned by imagesService which contains a host.
```
Comment (alexcepoi) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:194]: This sounds weird. Is it clear to you why that would be the case?

**Review by alexcepoi — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (anamariamilcu) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:91]: I understand, I will do like that from now on

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (anamariamilcu) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:174]: Well I checked before if the blobKeys array is empty.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (anamariamilcu) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Fixed

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (anamariamilcu) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:194]: Yes, we only need the relative path of the image, for it to be added(concatenated) in the url of my own page, I don't need the whole url in which the image is actually stored in the Blobstore service.

**Pull request merged**